dbms/src/Functions/FunctionsDateTime.h

```diff
@@ -1570,6 +1570,535 @@ class FunctionTimeSlots : public IFunction
+                            ". Should be a date or a date with time", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        if (!WhichDataType(arguments[1].type).isString())
+            throw Exception("Illegal type " + arguments[1].type->getName() + " of 2 argument of function " + getName() + ". Must be String.",
+                            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        if (arguments.size() == 3)
+        {
+            if (!WhichDataType(arguments[2].type).isString())
+                throw Exception("Illegal type " + arguments[2].type->getName() + " of 3 argument of function " + getName() + ". Must be String.",
+                                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
+    {
+        auto times = checkAndGetColumn<ColumnUInt32>(block.getByPosition(arguments[0]).column.get());
+        auto const_times = checkAndGetColumnConst<ColumnUInt32>(block.getByPosition(arguments[0]).column.get());
+
+        if (const_times)
+            throw Exception("Constant times are not supported yet");
+
+        const ColumnConst * pattern_column = checkAndGetColumnConst<ColumnString>(block.getByPosition(arguments[1]).column.get());
+
+        if (!pattern_column)
+            throw Exception("Illegal column " + block.getByPosition(arguments[1]).column->getName()
+                            + " of second ('format') argument of function " + getName()
+                            + ". Must be constant string.",
+                            ErrorCodes::ILLEGAL_COLUMN);
+
+        String pattern = pattern_column->getValue<String>();
+
+        std::vector<FormattingOperation> instructions = {};
+        size_t result_length = parsePattern(pattern, instructions);
+
+        const DateLUTImpl * time_zone_tmp = nullptr;
+        if (arguments.size() == 3)
+            time_zone_tmp = &extractTimeZoneFromFunctionArguments(block, arguments, 2, 0);
+        else
+            time_zone_tmp = &DateLUT::instance();
+
+        const DateLUTImpl & time_zone = *time_zone_tmp;
+
+        const ColumnUInt32::Container & vec = times->getData();
+
+        auto col_res = ColumnString::create();
+        auto & dst_data = col_res->getChars();
+        auto & dst_offsets = col_res->getOffsets();
+        dst_data.resize(vec.size() * (result_length + 1));
+        dst_offsets.resize(vec.size());
+
+        auto begin = reinterpret_cast<char *>(dst_data.data());
+        auto pos = begin;
+
+        for (size_t i = 0; i < vec.size(); ++i)
+        {
+            for(auto & instruction : instructions) {
+                instruction.action(pos, vec[i], time_zone);
+            }
+            *pos++ = '\0';
+            dst_offsets[i] = pos - begin;
+        }
+
```

Comment:
dst_data should be resized to the last offset.

Reply:
Correct me, please, if I am wrong,  but it is already resized, cause length of formatted string is known after parse phase.
But I've added it, anyway :)

Reply:
Is the exact length is known or just an upper bound?
If all formatters have fixed length - that that's perfect!

Reply:
That case, I consider the work on this task being complete - code, tests and docs are ready.

Reply:
Sorry, missed to answer - exact bound is known.